### ISSUE TITLE
higher for Reinforcement Learning?

### ISSUE THREAD

**Issue by rickyloynd-microsoft** [facebookresearch/higher#114]
Thank you for this library.

Consider an RL setting where the module passed to higher.innerloop_ctx() contains around 10M trainable parameters, and the unroll length is 32 timesteps, in a training loop that will run for 32M timesteps total. As I understand it, even though the structure of the module never changes, higher.innerloop_ctx() would need to be called one million times during training. Is higher performant enough for that? Or (ideally) is there a way to reset fmodel.parameters() (using fast-weight values from the end of the previous time window) without having to call higher.innerloop_ctx() on each time window?

**Comment by egrefen** [comment]
Thanks!

I think your biggest issue there is going to be the fact you need to hold something like 32x the forward and backward memory footprint of your model during inner loop unrolls. I haven't profiled things, but I would be surprised if the creation of `higher.innerloop_ctx()` was a significant cost relative to everything else that's happening.

If you profile this, and find that it is at the root of your issue, I would flag it here, but my guess would be it's better to call it each time you begin an inner loop to (hopefully) avoid something going wrong.

**Comment by rickyloynd-microsoft** [comment]
Thanks for the quick response. As you recommended for the RL scenario above, my code is calling higher.innerloop_ctx() at the beginning of each inner rollout window. I'm passing the Adam optimizer to higher.innerloop_ctx(), so a fresh copy of DifferentiableAdam gets instantiated each time, which means that Adam's internal state is not carried over between rollout windows. This loss of internal state seems to discard the advantages of Adam over SGD. Is there a recommended way for me to avoid this problem by somehow passing the optimizer's internal state into the next instantiation of the optimizer?
